src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl

```diff
@@ -16,19 +16,104 @@ end
   return all(x->radical_membership(x, defining_ideal(P)), gens(ambient_coordinate_ring(P)))
 end
 
+@doc raw"""
+    function is_smooth(P::AbsProjectiveScheme; algorithm=:default) -> Bool
 
-# TODO: Jacobi criterion
-function is_smooth(P::AbsProjectiveScheme; algorithm=:covered)
+Check whether the scheme `P` is smooth.
+
+# Algorithms
+
+There are three possible algorithms for checking smoothness, determined
+by the value of the keyword argument `algorithm`:
+  * `:projective_jacobian` - uses the Jacobian criterion for projective
+    varieties, see Exercise 4.2.10 of [Liu06](@cite),
+  * `:covered` - converts to a covered scheme,
+  * `:affine_cone` - checks that the affine cone is smooth outside the origin.
+
+The `:projective_jacobian` and the `:covered` algorithms first check
+that the scheme is equidimensional, which can be expensive.
+
+By default, if the base ring is a field and the scheme has already been
+computed to be equidimensional, then the `:projective_jacobian`
+algorithm is used. Otherwise, the `:affine_cone` algorithm is used.
+
+# Examples
+```jldoctest
+julia> A, (x, y, z) = grade(QQ["x", "y", "z"][1]);
+
+julia> B, _ = quo(A, ideal(A, [x^2 + y^2]));
+
+julia> C = projective_scheme(B)
+Projective scheme
+  over rational field
+defined by ideal(x^2 + y^2)
+
+julia> is_smooth(C)
+false
+
+julia> is_smooth(C; algorithm=:covered_jacobian)
+false
+```
+"""
+function is_smooth(P::AbsProjectiveScheme; algorithm=:default)
   get_attribute!(P, :is_smooth) do
-    if algorithm == :covered
-      return is_smooth(covered_scheme(P))
+    if algorithm == :default
+      if (
+        base_ring(P) isa Field
+        && has_attribute(P, :is_equidimensional)
+        && is_equidimensional(P)
+      )
+        algorithm = :projective_jacobian
+      else
+        algorithm = :affine_cone
+      end
     end
-    if algorithm == :jacobi
-      error("jacobi criterion not implemented")
+
+    algorithms = [
+      :projective_jacobian,
+      :covered_jacobian,
+      :affine_cone,
+    ]
+    if !(algorithm in algorithms)
+      throw(ArgumentError(
+        "the optional argument to the function is_smooth can only be one"
+         + " of the following: " + join(algorithms, ", ") + "."
+      ))
+    end
+
+    if algorithm == :covered_jacobian
+      return _jacobian_criterion(covered_scheme(P))
+    elseif algorithm == :affine_cone
+      aff = affine_cone(P)[1]
```

Comment:
An equivalent way to write this is
```suggestion
      aff, _ = affine_cone(P)
```
Personally, I have no preference.

Reply:
Edited according to your suggestion.